### ISSUE TITLE
Upgrade to django 2.0

### ISSUE THREAD

**Issue by chris48s** [DemocracyClub/UK-Polling-Stations#1092]
Django 1.8 is an LTS release and still gets security patches but it is EOL in April 2018 - see https://www.djangoproject.com/download/

We need to sort out how to migrate at some point but we are definitely relying on some functionality which is deprecated in more recent versions so it is a non-trivial task.

There is also the question of whether we should target 1.11 or wait a bit and try to jump straight to 2.0?

@symroe - probs worth having a chat about this one at some point.

**Comment by party98** [comment]
Is this up for grabs?

**Comment by party98** [comment]
It seems that 2.0 is available now

**Comment by chris48s** [comment]
Hi @party98 Thanks for showing interest in this. You're right - Django 2.0 was released yeterday.

In terms of this project, we've yet to decide on the migration path.

I've not reviewd the release notes for 2.0 yet, but at the moment I'm kind of leaning towards 1.11. It is a LTS release and it will probably be a lower bar to migrate the actual application to 1.11 (i.e: less non-BC changes).

Additionally we rely on a number of django-based dependencies (either third party, or which we maintain ourselves - see https://github.com/DemocracyClub/UK-Polling-Stations/blob/master/requirements/base.txt ). Most or all of them are tested under 1.11. If we want to go to 2.0 we would need to be in a place where we know that all of our dependencies are Django 2.0-ready and we are definitely not there right now.

In terms of the work needed to migrate our application, its not just a case of bumping the version and running the tests. As a minimum, we'll need to:

* Review all our dependencies for compatibility - if there are compatibility blockers, we may need to update packages we maintain ourselves (e.g: https://github.com/DemocracyClub/dc_base_theme , https://github.com/DemocracyClub/dc_signup_form , https://github.com/chris48s/django-apiblueprint-view ) or wait for/submit PRs to/switch out third-party packages with compatibility issues.
* Make code updates based on known deprecations/BC-breaks (e.g: `from django.core.urlresolvers import reverse` --> `from django.urls import reverse`, update `SessionMiddleware` + so on)
* The way we are providing the embedded and whitelabel versions of our site rely on a deprecated use of `django.conf.urls.patterns()` which is removed in Django 1.10 onwards. We will need to work out how to implement this functionality in a newer version of the framework.
* While we do have good test coverage, there is a bunch of code that is not under test (see https://coveralls.io/github/DemocracyClub/UK-Polling-Stations ) which will need manual review for compatibility.

Basically, if you're looking for an issue where you can bump a version number, make sure the tests go green, submit a PR and move on, that is not what this is: It is going to require some input from someone familiar with the application.

If you have expertise in upgrading legacy django applications that you are looking to help us with, we'd be grateful for that. Maybe a good place to start would be to provide some input or thoughts on some of the above points.

**Comment by chris48s** [comment]
Reviewing the changelog in https://github.com/DemocracyClub/UK-Polling-Stations/pull/1106 I note that `django-extensions` doesn't yet declare full Django 2.0 support

**Comment by chris48s** [comment]
I've done a bit of a first pass and identified some change we need to make regardless of whether we upgrade to 1.11 or 2.0:

Settings:

```
RemovedInDjango110Warning: Session verification will become mandatory in Django 1.10.
Please add 'django.contrib.auth.middleware.SessionAuthenticationMiddleware' to your
MIDDLEWARE_CLASSES setting when you are ready to opt-in after reading the upgrade
considerations in the 1.8 release notes.
```

```
RemovedInDjango110Warning: You haven't defined a TEMPLATES setting.
You must do so before upgrading to Django 1.10.
Otherwise Django will be unable to load templates.
```

```
RemovedInDjango110Warning: django.core.context_processors is deprecated
in favor of django.template.context_processors.
```

Code:

```
polling_stations/apps/data_finder/views.py
RemovedInDjango110Warning: `data_finder.views.AddressFormView.get_form` method
must define a default value for its `form_class` argument.
```

```
urls.py (review all apps)
RemovedInDjango110Warning: django.conf.urls.patterns() is deprecated
and will be removed in Django 1.10.
Update your urlpatterns to be a list of django.conf.urls.url() instances instead.

RemovedInDjango110Warning: Support for string view arguments to url() is deprecated
and will be removed in Django 1.10. Pass the callable instead.
```

That's probably not exhaustive, but fixing those things is a necessary step on the road to any upgrade, so fixing those seems like a good place to start on this.. then we can see where we are

**Comment by chris48s** [comment]
Upgrade to Django 1.11 now in progress in #1124

Further work
* Continue testing that and get it merged + deployed
* Review what else we need to do to update to 2.0 from there:
    * Fix 1.11 to 2.0 deprecations (there are a non-zero number of these)
    * Finish https://github.com/DemocracyClub/dc_signup_form/pull/3
    * Review `dc_base_theme` for compatibility
    * Review 3rd party packages for any compatibility issues

**Comment by chris48s** [comment]
going to bin this issue for now. We'll address the next upgrade when we take all of our sites from 1.11 --> 2.2 post-May
